### PR TITLE
Plot sparse data

### LINKED ISSUE #629: Plotting events
We need a way to visualize event (sparse) data. This can be done with scatter plots and histogramming.

### PR DESCRIPTION
**Pull request by nvaytet**
This adds the possibility to plot sparse data (neutron events) directly without the need for histogramming the data first. Inidividual data points (events) can be plotted as a scatter plot, or histogrammed on the fly into a bar/image plot.

Example:
```Python
N = 10
M = 50
var = sc.Variable(dims=[Dim.X, Dim.Tof],
                  shape=[N, sc.Dimensions.Sparse])
for i in range(N):
    var[Dim.X, i].values = np.random.normal(50.0, scale=20.0, size=int(np.random.rand()*M))

d = sc.Dataset()
d.coords[Dim.X] = sc.Variable([Dim.X], values=np.arange(N, dtype=np.float64))
d['a'] = sc.DataArray(coords={Dim.Tof: var})
```
Plotting individual events:
```Python
plot(d)
```
Screenshot at 2019-10-24 22-10-42
Plotting as image:
```Python
plot(d, bins=50)
```
Screenshot at 2019-10-24 22-12-22

If weights are present in the sparse data, they are represented by colors:
```Python
d['a'] = sc.DataArray(data=var, coords={Dim.Tof: var})
plot(d)
```
Screenshot at 2019-10-24 22-18-02

Fixes #629 

**Additional notes:**

- The input for 1d, 2d, 3d plotting was uniformized so that all functions take in a `Dataset` as an input (before it was either lists of `DataArray` or `DataArray`). As a consequence, it is no longer possible to do 1D plots with `plot([d1, d2])` where `d1` and `d2` are `Dataset`s, you can only plot the contents of one `Dataset` at the same time, but it just seemed strange that the 1D plotting was a special case, the only one you could do this with. The `plotting.ipynb` user-guide notebook was updated as a result.

-

### CODE REVIEW

**Comment by nvaytet** [comment]
> Looks good and notebook worked as described.
> 
> Not related to the changes here, but in the plotting notebook, you refer to collapsing dimensions, but it's not clear what the scipp plotting does with the data over the collapsed dimension, integrated - sampled?

Yeah, it's not super clear, and I don't even know how useful it is. It basically creates one 1D plot per element in the other dimensions. So if you have 2D data of size `{Dim.X: 5, Dim.Tof: 100}`, and you want to plot the individual spectra, you can use `plot(d, collapse=Dim.Tof)` to create a 1D plot with 5 lines/histograms each with 100 data points with `Dim.Tof` along the x axis.

If you have `{Dim.Y: 6, Dim.X: 5, Dim.Tof: 100}`, then `plot(d, collapse=Dim.Tof)` will create a 1D plot with 30 different lines/histograms. So the data is not added up or anything. @SimonHeybrock also rightly pointed out earlier that `collapse` was not a great name because `plot(d, collapse=Dim.Tof)` would suggest this is the dimension we are collapsing is `Dim.Tof` while in fact this is the dimension we are retaining.

So maybe we should just get rid of it, or find better names, or make it do something else like actually collapse/flatten the data along one dimension, if that is something people do. But then it should probably be an in-built `scipp` method like `sum` or `mean`, or `group`, and not limited to plotting.
It was originally inspired by the Mantid "Waterfall plots"
wplot
but I did not like the fact that the values on the x and y axes only really applied to the first curve

**Comment by OwenArnold** [comment]
> > Looks good and notebook worked as described.
> > Not related to the changes here, but in the plotting notebook, you refer to collapsing dimensions, but it's not clear what the scipp plotting does with the data over the collapsed dimension, integrated - sampled?
> 
> Yeah, it's not super clear, and I don't even know how useful it is. It basically creates one 1D plot per element in the other dimensions. So if you have 2D data of size `{Dim.X: 5, Dim.Tof: 100}`, and you want to plot the individual spectra, you can use `plot(d, collapse=Dim.Tof)` to create a 1D plot with 5 lines/histograms each with 100 data points with `Dim.Tof` along the x axis.
> 
> If you have `{Dim.Y: 6, Dim.X: 5, Dim.Tof: 100}`, then `plot(d, collapse=Dim.Tof)` will create a 1D plot with 30 different lines/histograms. So the data is not added up or anything. @SimonHeybrock also rightly pointed out earlier that `collapse` was not a great name because `plot(d, collapse=Dim.Tof)` would suggest this is the dimension we are collapsing is `Dim.Tof` while in fact this is the dimension we are retaining.
> 
> So maybe we should just get rid of it, or find better names, or make it do something else like actually collapse/flatten the data along one dimension, if that is something people do. But then it should probably be an in-built `scipp` method like `sum` or `mean`, or `group`, and not limited to plotting.
> It was originally inspired by the Mantid "Waterfall plots"
> wplot
> but I did not like the fact that the values on the x and y axes only really applied to the first curve

Waterfall plots are apparently something that neutron scattering users like a lot. I would opt for better names rather than removing the feature.

File: python/src/scipp/plot/sparse.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        return xmin, xmax
+
+
+def histogram_sparse_data(input_data, sparse_dim, bins):
+    var = next(iter(input_data))[1]
+    if isinstance(bins, bool):
+        bins = 256
+    if isinstance(bins, int):
+        # Find min and max
+        xmin, xmax = visit_sparse_data(input_data, sparse_dim)
```
Comment (SimonHeybrock) [inline python/src/scipp/plot/sparse.py]: Is this a performance issue with many events? We probably want something like this on the C++ side.

Comment (nvaytet) [reply python/src/scipp/plot/sparse.py]: Yes, something on the c++ side would be optimal. Also, can we use the new `counts` function you added to make the `show` diagrams?

**Pull request merged**